### PR TITLE
Vararray defaults

### PR DESCRIPTION
**Pull request by dheckenberg**
This changeset is to properly handle defaults for vararray parameters.  Without the patch, the following cases won't won't work/may crash if the vararray initializer has length > 1:
- oslinfo   (see testsuite/oslinfo-arrayparams)
- OSL arraylength() (see testsuite/vararray-connect, testsuite/vararray-param)
- OSL array access (see testsuite/vararray-default)

I also tightened up the parser grammar to error if a vararray output is specified (see testsuite/oslc-err-outputparamvararray).  This should probably be noted explicitly in the language spec.

A small patch it also required to OIIO for the TypeDesc struct.

### CODE REVIEW

File: src/include/osl_pvt.h [hunk 1]
```diff
@@ -205,7 +205,15 @@ class TypeSpec {
 
     /// Returns the length of the array, or 0 if not an array.
     ///
-    int arraylength () const { return m_simple.arraylen; }
+    int arraylength () const { return abs(m_simple.arraylen); }
```
Comment (fpsunflower) [inline src/include/osl_pvt.h]: Per my comments on the OSL side -- would it make sense to instead keep the OIIO semantics the same, but add a field in here to indicate the default varlen size?

**Comment by fpsunflower** [comment]
LGTM to the changes, aside from the discussion on where the concept of "default" length really belongs. This might just be a matter of extra docs or comments to clarify my mental model.

**Comment by lgritz** [comment]
See #497 for my proposed alternate patch (which is still mostly Daniel's work, but I found a set of changes that let it be entirely self-contained in OSL land without any need for modifications to OIIO or semantic-bending use of TypeDesc).

**Comment by lgritz** [comment]
Closing -- this was superseded by #497

**Pull request closed without merging**